From 898e9ba56b85f8ff3aff3359c1e9f247899a6db8 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 29 Oct 2022 17:58:28 +0200
Subject: [PATCH 0771/1013] evl/thread: add deferred wchan requeuing

Changing the priority and/or scheduling policy a thread undergoes
requires to requeue it in the wait channel it may be pending on at the
time of that change. In turn, such requeuing may require to walk the
PI chain. Meanwhile, we have to abide by the common locking order
which is stated as:

      lock(wchan)
        lock(thread)
          lock(rq)

Any other order would certainly trigger an ABBA locking issue. For
this reason, we cannot have evl_set_thread_schedparam_locked() perform
such requeuing immediately, only someone else in the call stack may do
so later when the target thread and rq locks have been dropped.

To address this, we introduce the T_WCHAN flag to mark a thread which
needs a deferred requeuing after its (weighted) scheduling priority
has changed, handling it next wherever applicable at the first
opportunity, by calling evl_adjust_wait_priority().

CAUTION: to prevent any priority inversion, all code between the
moment T_CHAN is set and the subsequent call to
evl_adjust_wait_priority() must run with irqs off.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/mutex.h       |  10 --
 include/evl/sched.h       |  25 ++++
 include/evl/thread.h      |  20 +++
 include/evl/wait.h        |  14 +++
 include/uapi/evl/thread.h |   2 +-
 kernel/evl/mutex.c        | 259 ++------------------------------------
 kernel/evl/sched/core.c   |  40 +++++-
 kernel/evl/thread.c       |  43 ++++---
 kernel/evl/wait.c         | 218 ++++++++++++++++++++++++++++++++
 9 files changed, 352 insertions(+), 279 deletions(-)

diff --git a/include/evl/mutex.h b/include/evl/mutex.h
index 3be0f7e81c83..045db8db328b 100644
--- a/include/evl/mutex.h
+++ b/include/evl/mutex.h
@@ -30,13 +30,6 @@ struct evl_thread;
 /* Counted in thread->held_mutex_count. */
 #define EVL_MUTEX_COUNTED	BIT(4)
 
-/* Modes for PI chain walk. */
-enum evl_walk_mode {
-	evl_pi_adjust,		/* Adjust priority of members. */
-	evl_pi_reset,		/* Revert members to their base priority. */
-	evl_pi_check,		/* Check the PI chain (no change). */
-};
-
 struct evl_mutex {
 	int wprio;
 	int flags;
@@ -156,7 +149,4 @@ void evl_unlock_kmutex(struct evl_kmutex *kmutex)
 	return evl_unlock_mutex(&kmutex->mutex);
 }
 
-void evl_adjust_wait_priority(struct evl_thread *thread,
-			      enum evl_walk_mode mode);
-
 #endif /* !_EVL_MUTEX_H */
diff --git a/include/evl/sched.h b/include/evl/sched.h
index 279ef6442b43..7bba6903c6ed 100644
--- a/include/evl/sched.h
+++ b/include/evl/sched.h
@@ -386,6 +386,12 @@ static inline bool evl_cannot_block(void)
 		raw_spin_unlock_irqrestore(&(__thread)->lock, __flags);	\
 	} while (0)
 
+#define evl_put_thread_rq_check(__thread, __rq, __flags)		\
+	do {								\
+		evl_put_thread_rq_check_noirq(__thread, __rq);		\
+		hard_local_irq_restore(__flags);			\
+	} while (0)
+
 #define evl_get_thread_rq_noirq(__thread)				\
 	({								\
 		struct evl_rq *__rq;					\
@@ -401,6 +407,25 @@ static inline bool evl_cannot_block(void)
 		raw_spin_unlock(&(__thread)->lock);			\
 	} while (0)
 
+/*
+ * Unpin the thread, checking for a pending requeue operation in a
+ * wait channel. See comment in drop_boosters() for details about the
+ * requirements to achieve correctness in this case.
+ *
+ * CAUTION: we should hold NO lock on entry to this macro, at the very
+ * least we must not hold any wchan or thread lock, otherwise an ABBA
+ * issue is certain.
+ */
+#define evl_put_thread_rq_check_noirq(__thread, __rq)			\
+	do {								\
+		bool __need_requeue = (__thread)->info & T_WCHAN;	\
+		if (__need_requeue)					\
+			(__thread)->info &= ~T_WCHAN;			\
+		evl_put_thread_rq_noirq(__thread, __rq);		\
+		if (__need_requeue)					\
+			evl_adjust_wait_priority(__thread, evl_pi_adjust); \
+	} while (0)
+
 bool evl_set_effective_thread_priority(struct evl_thread *thread,
 				       int prio);
 
diff --git a/include/evl/thread.h b/include/evl/thread.h
index f2f2f09ff46d..24350bf9780b 100644
--- a/include/evl/thread.h
+++ b/include/evl/thread.h
@@ -21,6 +21,7 @@
 #include <evl/timer.h>
 #include <evl/sched/param.h>
 #include <evl/factory.h>
+#include <evl/assert.h>
 #include <uapi/evl/thread.h>
 #include <uapi/evl/signal.h>
 #include <uapi/evl/sched.h>
@@ -188,6 +189,25 @@ void evl_set_sync_uwindow(struct evl_thread *curr, int state_bits)
 	}
 }
 
+static inline
+void evl_double_thread_lock(struct evl_thread *t1, struct evl_thread *t2)
+{
+	EVL_WARN_ON_ONCE(CORE, t1 == t2);
+	EVL_WARN_ON_ONCE(CORE, !hard_irqs_disabled());
+
+	/*
+	 * Prevent ABBA deadlock, always lock threads in address
+	 * order. The caller guarantees t1 and t2 are distinct.
+	 */
+	if (t1 < t2) {
+		raw_spin_lock(&t1->lock);
+		raw_spin_lock(&t2->lock);
+	} else {
+		raw_spin_lock(&t2->lock);
+		raw_spin_lock(&t1->lock);
+	}
+}
+
 void __evl_test_cancel(struct evl_thread *curr);
 
 void evl_discard_thread(struct evl_thread *thread);
diff --git a/include/evl/wait.h b/include/evl/wait.h
index 6667b786d4d5..d1c1f92966e5 100644
--- a/include/evl/wait.h
+++ b/include/evl/wait.h
@@ -31,6 +31,13 @@ struct evl_wait_channel {
 	const char *name;
 };
 
+/* Modes for PI chain walk. */
+enum evl_walk_mode {
+	evl_pi_adjust,		/* Adjust priority of members. */
+	evl_pi_reset,		/* Revert members to their base priority. */
+	evl_pi_check,		/* Check the PI chain (no change). */
+};
+
 #ifdef CONFIG_PROVE_LOCKING
 struct evl_lock_key_addr {
 	struct lock_class_key *addr;
@@ -161,4 +168,11 @@ int evl_reorder_wait(struct evl_thread *waiter,
 void evl_requeue_wait(struct evl_wait_channel *wchan,
 		struct evl_thread *waiter);
 
+void evl_adjust_wait_priority(struct evl_thread *thread,
+			      enum evl_walk_mode mode);
+
+int evl_walk_pi_chain(struct evl_wait_channel *orig_wchan,
+		struct evl_thread *orig_waiter,
+		enum evl_walk_mode mode);
+
 #endif /* !_EVL_WAIT_H_ */
diff --git a/include/uapi/evl/thread.h b/include/uapi/evl/thread.h
index 8aec520a6652..2dbe7f77fef2 100644
--- a/include/uapi/evl/thread.h
+++ b/include/uapi/evl/thread.h
@@ -44,7 +44,7 @@
 #define T_RMID    0x00000002 /* Pending on a removed resource */
 #define T_BREAK   0x00000004 /* Forcibly awaken from a wait state */
 #define T_KICKED  0x00000008 /* Forced out of OOB context */
-/* free: 0x00000010 */
+#define T_WCHAN   0x00000010 /* Need to requeue in wait channel */
 /* free: 0x00000020 */
 #define T_CANCELD 0x00000040 /* Cancellation request is pending */
 #define T_PIALERT 0x00000080 /* Priority inversion alert (HM notified) */
diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index 6808e1eccd18..02ab80c38010 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -19,25 +19,6 @@
 #define for_each_evl_mutex_waiter(__pos, __mutex)			\
 	list_for_each_entry(__pos, &(__mutex)->wchan.wait_list, wait_next)
 
-static inline
-void double_thread_lock(struct evl_thread *t1, struct evl_thread *t2)
-{
-	EVL_WARN_ON_ONCE(CORE, t1 == t2);
-	EVL_WARN_ON_ONCE(CORE, !hard_irqs_disabled());
-
-	/*
-	 * Prevent ABBA deadlock, always lock threads in address
-	 * order. The caller guarantees t1 and t2 are distinct.
-	 */
-	if (t1 < t2) {
-		raw_spin_lock(&t1->lock);
-		raw_spin_lock(&t2->lock);
-	} else {
-		raw_spin_lock(&t2->lock);
-		raw_spin_lock(&t1->lock);
-	}
-}
-
 static inline int get_ceiling_value(struct evl_mutex *mutex)
 {
 	/*
@@ -71,229 +52,11 @@ void enable_inband_switch(struct evl_thread *curr, struct evl_mutex *mutex)
 		mutex->flags &= ~EVL_MUTEX_COUNTED;
 		if (atomic_dec_return(&curr->held_mutex_count) < 0) {
 			atomic_set(&curr->held_mutex_count, 0);
-			EVL_WARN_ON_ONCE(CORE, 1);
+			EVL_WARN_ON(CORE, 1);
 		}
 	}
 }
 
-/*
- * Walking the PI chain deals with the following items:
- *
- * waiter := the thread which waits on wchan
- * wchan := the wait channel being processed
- * owner := the current owner of wchan
- *
- * In order to go fine-grained and escape ABBA situations, we hold
- * three locks at most while walking the PI chain, keeping a
- * consistent locking sequence among them as follows:
- *
- * lock(wchan)
- *      lock(waiter, owner)
- *
- * Rules:
- *
- * 1. holding wchan->lock guarantees that wchan->owner is stable.
- *
- * 2. holding thread->lock guarantees that thread->wchan is stable.
- *
- * 3. there is no rule #3.
- *
- * On entry:
- * orig_wchan->lock held + orig_waiter->lock held, irqs off.
- */
-static int walk_pi_chain(struct evl_wait_channel *orig_wchan,
-			struct evl_thread *orig_waiter,
-			enum evl_walk_mode mode)
-{
-	static atomic64_t pi_serial;
-	struct evl_thread *waiter = orig_waiter, *curr = evl_current(), *owner;
-	struct evl_wait_channel *wchan = orig_wchan, *next_wchan;
-	s64 serial = atomic64_inc_return(&pi_serial);
-	bool do_reorder = false;
-	int ret = 0;
-
-	assert_hard_lock(&orig_wchan->lock);
-	assert_hard_lock(&orig_waiter->lock);
-
-	/*
-	 * The loop is always exited with two, and only two locks
-	 * held, namely: wchan->lock and waiter->lock. Also, *wchan
-	 * advances in lock-step with *waiter.
-	 *
-	 * Since we drop locks inside the PI walk, we might have
-	 * concurrent walks on the same chain from different CPUs,
-	 * contradicting each other when it comes to the priority to
-	 * set for a thread. We solve this issue as follows:
-	 *
-	 * - a unique serial number is drawn at each invocation of
-	 * this routine (pi_serial). A wait channel maintains a cached
-	 * copy of this serial number, which is updated under lock.
-	 *
-	 * - before updating the priority of the owner of a wait
-	 * channel, the serial number drawn on entry (pi_serial) is
-	 * compared to the cached value: if this value is lower than
-	 * pi_serial, the change to wchan->owner can proceed, and the
-	 * cached copy is updated into the wait channel. Otherwise,
-	 * this means that a concurrent walk started later must
-	 * override the effect of the current one, in which case the
-	 * current PI walk is aborted.
-	 *
-	 * IOW, we ensure that the latest PI walk always overrides the
-	 * effect of walks started earlier. The initial serialization
-	 * is granted by orig_wchan->lock, which is held on entry to
-	 * this routine.
-	 */
-	for (;;) {
-		owner = wchan->owner;
-		if (!owner) /* End of PI chain, we are done. */
-			break;
-
-		if (owner == orig_waiter) {
-			ret = -EDEADLK;
-			break;
-		}
-
-		/*
-		 * Lock both the waiter and owner by increasing
-		 * address order to escape the ABBA issue.  To this
-		 * end, we have to drop waiter->lock temporarily
-		 * first, so that we can re-lock it at the right place
-		 * in the sequence. In the meantime, the waiter is
-		 * prevented to go stale by holding a reference on it.
-		 */
-		evl_get_element(&waiter->element);
-		raw_spin_unlock(&waiter->lock);
-
-		double_thread_lock(waiter, owner);
-		evl_put_element(&waiter->element);
-
-		/*
-		 * Multiple PI walks may happen concurrently, detect
-		 * and abort the oldest ones based on their serial
-		 * number, ensuring the latest overrides them.
-		 */
-		if (mode != evl_pi_check && serial - wchan->pi_serial < 0)
-			break;
-
-		wchan->pi_serial = serial;
-
-		/*
-		 * Make sure the waiter did not stop waiting on wchan
-		 * while unlocked (we have been holding wchan->lock
-		 * across the unlocked section, so we know that wchan
-		 * did not go stale). If the waiter is current, it is
-		 * certainly about to block on wchan, so this check
-		 * does not apply (besides, waiter->wchan does not
-		 * point at wchan yet).
-		 */
-		if (waiter != curr && waiter->wchan != wchan)
-			break;
-
-		/*
-		 * The current wait channel may need to reorder its
-		 * internal wait queue due to a priority change for
-		 * the waiter during the previous iteration.
-		 *
-		 * This also means that for a short while, a waiter
-		 * may benefit from a priority boost not yet accounted
-		 * for in the wait queue of the channel its pends on,
-		 * which is ok since there cannot be any rescheduling
-		 * on the current CPU between both events.
-		 */
-		if (do_reorder) {
-			wchan->requeue_wait(wchan, waiter);
-			do_reorder = false;
-		}
-
-		if (mode == evl_pi_adjust) {
-			/*
-			 * If priorities already match, there is no
-			 * more change to propagate downstream. Stop
-			 * the walk now.
-			 */
-			if (waiter->wprio == owner->wprio) {
-				raw_spin_unlock(&owner->lock);
-				break;
-			}
-			/* owner.cprio <- waiter.cprio */
-			evl_track_thread_policy(owner, waiter);
-			do_reorder = true;
-		} else if (mode == evl_pi_reset) {
-			/* owner.cprio <- owner.bprio */
-			if (owner->state & T_BOOST)
-				evl_track_thread_policy(owner, owner);
-			do_reorder = true;
-		}
-
-		/*
-		 * Drop the lock on the current wait channel we don't
-		 * need anymore, so that we don't risk entering an
-		 * ABBA pattern lockdep would complain about as a
-		 * result of locking owner->wchan in
-		 * evl_get_thread_wchan().
-		 */
-		raw_spin_unlock(&waiter->lock);
-		raw_spin_unlock(&wchan->lock);
-		/* Now acquire owner->wchan if non-NULL. */
-		next_wchan = evl_get_thread_wchan(owner);
-		if (!next_wchan) {
-			raw_spin_unlock(&owner->lock);
-			raw_spin_lock(&orig_wchan->lock);
-			raw_spin_lock(&orig_waiter->lock);
-			return 0;
-		}
-
-		waiter = owner;
-		wchan = next_wchan;
-	}
-
-	if (waiter != orig_waiter) {
-		raw_spin_unlock(&waiter->lock);
-		raw_spin_unlock(&wchan->lock);
-		raw_spin_lock(&orig_wchan->lock);
-		raw_spin_lock(&orig_waiter->lock);
-	}
-
-	return ret;
-}
-
-/*
- * Requeue a thread which changed priority in the wait channel it
- * pends on, then propagate the change down the PI chain.
- *
- * On entry: irqs off.
- */
-void evl_adjust_wait_priority(struct evl_thread *thread,
-			      enum evl_walk_mode mode)
-{
-	struct evl_wait_channel *wchan;
-	struct evl_thread *owner;
-
-	EVL_WARN_ON_ONCE(CORE, !hard_irqs_disabled());
-
-	raw_spin_lock(&thread->lock);
-	wchan = evl_get_thread_wchan(thread);
-	raw_spin_unlock(&thread->lock);
-
-	if (!wchan)
-		return;
-
-	owner = wchan->owner;
-	/* Careful: waitqueues have no owner. */
-	if (owner) {
-		double_thread_lock(thread, owner);
-		wchan->requeue_wait(wchan, thread);
-		raw_spin_unlock(&owner->lock);
-		walk_pi_chain(wchan, thread, mode);
-	} else {
-		raw_spin_lock(&thread->lock);
-		wchan->requeue_wait(wchan, thread);
-	}
-
-	raw_spin_unlock(&thread->lock);
-	evl_put_thread_wchan(wchan);
-}
-
 /* owner->lock + contender->lock held, irqs off */
 static int inherit_thread_priority(struct evl_thread *owner,
 				struct evl_thread *contender,
@@ -444,7 +207,7 @@ static void track_owner(struct evl_mutex *mutex,
 
 	assert_hard_lock(&mutex->wchan.lock);
 
-	if (EVL_WARN_ON_ONCE(CORE, prev != NULL))
+	if (EVL_WARN_ON(CORE, prev != NULL))
 		return;
 
 	raw_spin_lock(&owner->lock);
@@ -504,7 +267,7 @@ static void track_mutex_owner(struct evl_mutex *mutex, struct evl_thread *owner)
 	assert_hard_lock(&owner->lock);
 
 	if (prev) {
-		if (EVL_WARN_ON_ONCE(CORE, prev == owner))
+		if (EVL_WARN_ON(CORE, prev == owner))
 			return;
 		list_del(&mutex->next_owned);
 		evl_put_element(&prev->element);
@@ -644,7 +407,7 @@ static void adjust_owner_boost(struct evl_thread *owner)
 
 	raw_spin_lock(&owner->lock);
 
-	EVL_WARN_ON_ONCE(CORE, list_empty(&owner->boosters));
+	EVL_WARN_ON(CORE, list_empty(&owner->boosters));
 
 	/*
 	 * Fetch the mutex currently queuing the top waiter, among all
@@ -669,7 +432,7 @@ static void adjust_owner_boost(struct evl_thread *owner)
 		}
 		top_waiter = list_first_entry(&mutex->wchan.wait_list,
 					struct evl_thread, wait_next);
-		double_thread_lock(owner, top_waiter);
+		evl_double_thread_lock(owner, top_waiter);
 		/* Prevent spurious round-robin effect in runqueue. */
 		if (top_waiter->wprio != owner->wprio)
 			evl_track_thread_policy(owner, top_waiter);
@@ -863,7 +626,7 @@ void __evl_init_mutex(struct evl_mutex *mutex,
 
 	mutex->fastlock = fastlock;
 	atomic_set(fastlock, EVL_NO_HANDLE);
-	mutex->flags = type & ~(EVL_MUTEX_PIBOOST|EVL_MUTEX_CEILING);
+	mutex->flags = type;
 	mutex->wprio = -1;
 	mutex->ceiling_ref = ceiling_ref;
 	mutex->clock = clock;
@@ -1126,7 +889,7 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 	if (unlikely(curr->state & T_WOLI))
 		detect_inband_owner(mutex, curr);
 
-	double_thread_lock(curr, owner);
+	evl_double_thread_lock(curr, owner);
 
 	walk_mode = evl_pi_check;
 	if (mutex->flags & EVL_MUTEX_PI && curr->wprio > owner->wprio) {
@@ -1142,7 +905,7 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 
 	raw_spin_unlock(&owner->lock);
 
-	ret = walk_pi_chain(&mutex->wchan, curr, walk_mode);
+	ret = evl_walk_pi_chain(&mutex->wchan, curr, walk_mode);
 	assert_hard_lock(&mutex->wchan.lock);
 	assert_hard_lock(&curr->lock);
 	if (ret) {
@@ -1214,11 +977,11 @@ void __evl_unlock_mutex(struct evl_mutex *mutex)
 	 * reschedules the thread dropping it.
 	 */
 	if (mutex->flags & EVL_MUTEX_CEILING) {
-		EVL_WARN_ON_ONCE(CORE, mutex->flags & EVL_MUTEX_PIBOOST);
+		EVL_WARN_ON(CORE, mutex->flags & EVL_MUTEX_PIBOOST);
 		mutex->flags &= ~EVL_MUTEX_CEILING;
 		drop_booster(mutex);
 	} else if (mutex->flags & EVL_MUTEX_PIBOOST) {
-		EVL_WARN_ON_ONCE(CORE, mutex->flags & EVL_MUTEX_CEILING);
+		EVL_WARN_ON(CORE, mutex->flags & EVL_MUTEX_CEILING);
 		mutex->flags &= ~EVL_MUTEX_PIBOOST;
 		drop_booster(mutex);
 	}
@@ -1258,7 +1021,7 @@ void evl_unlock_mutex(struct evl_mutex *mutex)
 			if (curr->state & T_WOLI)
 				evl_notify_thread(curr, EVL_HMDIAG_LKIMBALANCE, evl_nil);
 		} else {
-			EVL_WARN_ON_ONCE(CORE, 1);
+			EVL_WARN_ON(CORE, 1);
 		}
 		return;
 	}
diff --git a/kernel/evl/sched/core.c b/kernel/evl/sched/core.c
index 191a94cf6c2b..01387e32b92d 100644
--- a/kernel/evl/sched/core.c
+++ b/kernel/evl/sched/core.c
@@ -264,6 +264,7 @@ void evl_double_rq_unlock(struct evl_rq *rq1, struct evl_rq *rq2)
 		raw_spin_unlock(&rq2->lock);
 }
 
+/* irqs off. */
 static void migrate_rq(struct evl_thread *thread, struct evl_rq *dst_rq)
 {
 	struct evl_sched_class *sched_class = thread->sched_class;
@@ -300,7 +301,15 @@ static void migrate_rq(struct evl_thread *thread, struct evl_rq *dst_rq)
 	evl_double_rq_unlock(src_rq, dst_rq);
 }
 
-/* thread->lock held, hard irqs off. @thread must be running in-band. */
+/*
+ * Move a thread to a different runqueue. Some sched_migrate() policy
+ * handlers might change the scheduling class and/or priority of the
+ * target thread, our callers MAY HAVE TO check for T_WCHAN to
+ * determine whether the wait channel it might pend on should be
+ * adjusted accordingly.
+ *
+ * thread->lock held, hard irqs off. @thread must be running in-band.
+ */
 void evl_migrate_thread(struct evl_thread *thread, struct evl_rq *dst_rq)
 {
 	assert_hard_lock(&thread->lock);
@@ -327,11 +336,14 @@ static void check_cpu_affinity(struct task_struct *p) /* inband, hard irqs off *
 	struct evl_thread *thread = evl_thread_from_task(p);
 	int cpu = task_cpu(p);
 	struct evl_rq *rq = evl_cpu_rq(cpu);
+	bool need_requeue = false;
 
 	raw_spin_lock(&thread->lock);
 
-	if (likely(rq == thread->rq))
-		goto out;
+	if (likely(rq == thread->rq)) {
+		raw_spin_unlock(&thread->lock);
+		return;
+	}
 
 	/*
 	 * Resync the EVL and in-band schedulers upon migration from
@@ -372,8 +384,22 @@ static void check_cpu_affinity(struct task_struct *p) /* inband, hard irqs off *
 	}
 
 	evl_migrate_thread(thread, rq);
-out:
+
+	/*
+	 * Check for a wait channel requeuing. Open code the portion
+	 * of the evl_put_thread_rq_check_noirq() logic we need.
+	 */
+	if (thread->info & T_WCHAN) {
+		raw_spin_lock(&thread->rq->lock);
+		thread->info &= ~T_WCHAN;
+		raw_spin_unlock(&thread->rq->lock);
+		need_requeue = true;
+	}
+
 	raw_spin_unlock(&thread->lock);
+
+	if (need_requeue)
+		evl_adjust_wait_priority(thread, evl_pi_adjust);
 }
 
 #else
@@ -956,6 +982,12 @@ void __evl_schedule(void) /* oob or/and hard irqs off (CPU migration-safe) */
 	 * a root thread never bears this bit.
 	 */
 	curr = this_rq->curr;
+	/*
+	 * Deferred WCHAN requeuing must be handled prior to
+	 * rescheduling.
+	 */
+	EVL_WARN_ON(CORE, curr->info & T_WCHAN);
+
 	if (curr->state & T_USER)
 		evl_commit_monitor_ceiling();
 
diff --git a/kernel/evl/thread.c b/kernel/evl/thread.c
index e43250186bf8..4a3c6b5bb941 100644
--- a/kernel/evl/thread.c
+++ b/kernel/evl/thread.c
@@ -152,6 +152,10 @@ static void pin_to_initial_cpu(struct evl_thread *thread)
 	 * of mapping the current in-band task to it. Therefore
 	 * evl_migrate_thread() can be called for pinning it on an
 	 * out-of-band CPU.
+	 *
+	 * NOTE: we do not need to check for T_WCHAN on return from
+	 * evl_migrate_thread(), there is no way the emerging thread
+	 * could be sleeping on a wait channel.
 	 */
 	rq = evl_cpu_rq(cpu);
 	raw_spin_lock_irqsave(&thread->lock, flags);
@@ -1048,7 +1052,7 @@ int evl_set_thread_schedparam(struct evl_thread *thread,
 
 	rq = evl_get_thread_rq(thread, flags);
 	ret = evl_set_thread_schedparam_locked(thread, sched_class, sched_param);
-	evl_put_thread_rq(thread, rq, flags);
+	evl_put_thread_rq_check(thread, rq, flags);
 
 	return ret;
 }
@@ -1063,8 +1067,8 @@ EXPORT_SYMBOL_GPL(evl_set_thread_schedparam);
  * On entry: thread->lock + thread->rq->lock held, irqs off.
  */
 int evl_set_thread_schedparam_locked(struct evl_thread *thread,
-				     struct evl_sched_class *sched_class,
-				     const union evl_sched_param *sched_param)
+				struct evl_sched_class *sched_class,
+				const union evl_sched_param *sched_param)
 {
 	int old_wprio, new_wprio, ret;
 
@@ -1079,13 +1083,18 @@ int evl_set_thread_schedparam_locked(struct evl_thread *thread,
 	new_wprio = thread->wprio;
 
 	/*
-	 * If the thread is sleeping on a wait channel, update its
-	 * position in the corresponding wait list, unless the
-	 * (weighted) priority has not changed (to prevent spurious
-	 * round-robin effects).
+	 * Only if the (weighted) priority actually changed - so that
+	 * we do not cause any spurious RR side effect - and the
+	 * thread is sleeping on a wait channel, tell the caller to
+	 * requeue it its wait list at the first opportunity. We
+	 * cannot do that here since this would trigger ABBA locking
+	 * issues with wchan->lock and/or thread->lock.
+	 *
+	 * CAVEAT: This must be done prior to rescheduling or
+	 * re-enabling irqs in order to prevent priority inversion.
 	 */
-	if (old_wprio != new_wprio && (thread->state & T_PEND))
-		thread->wchan->reorder_wait(thread, thread); /* FIXME: evl_adjust_wait_priority(thread), but we hold rq, BAD. */
+	if (old_wprio != new_wprio && thread->wchan)
+		thread->info |= T_WCHAN;
 
 	thread->info |= T_SCHEDP;
 	/* Ask the target thread to call back if in-band. */
@@ -1303,7 +1312,7 @@ void evl_demote_thread(struct evl_thread *thread)
 	sched_class = &evl_sched_weak;
 	evl_set_thread_schedparam_locked(thread, sched_class, &param);
 
-	evl_put_thread_rq(thread, rq, flags);
+	evl_put_thread_rq_check(thread, rq, flags);
 
 	/* Then unblock it from any wait state. */
 	evl_kick_thread(thread, 0);
@@ -2016,17 +2025,19 @@ static int set_sched_attrs(struct evl_thread *thread,
 	tslice = thread->rrperiod;
 	sched_class = evl_find_sched_class(&param, attrs, &tslice);
 	if (IS_ERR(sched_class)) {
-		ret = PTR_ERR(sched_class);
-		goto out;
+		evl_put_thread_rq(thread, rq, flags);
+		return PTR_ERR(sched_class);
 	}
 
 	ret = set_time_slice(thread, tslice);
-	if (ret)
-		goto out;
+	if (ret) {
+		evl_put_thread_rq(thread, rq, flags);
+		return ret;
+	}
 
 	ret = evl_set_thread_schedparam_locked(thread, sched_class, &param);
-out:
-	evl_put_thread_rq(thread, rq, flags);
+
+	evl_put_thread_rq_check(thread, rq, flags);
 
 	return ret;
 }
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index e00307e4a9f2..309b22bb958c 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -275,3 +275,221 @@ struct evl_thread *evl_wait_head(struct evl_wait_queue *wq)
 					struct evl_thread, wait_next);
 }
 EXPORT_SYMBOL_GPL(evl_wait_head);
+
+/*
+ * Requeue a thread which changed priority in the wait channel it
+ * pends on, then propagate the change down the PI chain.
+ *
+ * On entry: irqs off.
+ */
+void evl_adjust_wait_priority(struct evl_thread *thread,
+			      enum evl_walk_mode mode)
+{
+	struct evl_wait_channel *wchan;
+	struct evl_thread *owner;
+
+	EVL_WARN_ON_ONCE(CORE, !hard_irqs_disabled());
+
+	raw_spin_lock(&thread->lock);
+	wchan = evl_get_thread_wchan(thread);
+	raw_spin_unlock(&thread->lock);
+
+	if (!wchan)
+		return;
+
+	owner = wchan->owner;
+	/* Careful: waitqueues have no owner. */
+	if (owner) {
+		evl_double_thread_lock(thread, owner);
+		wchan->requeue_wait(wchan, thread);
+		raw_spin_unlock(&owner->lock);
+		evl_walk_pi_chain(wchan, thread, mode);
+	} else {
+		raw_spin_lock(&thread->lock);
+		wchan->requeue_wait(wchan, thread);
+	}
+
+	raw_spin_unlock(&thread->lock);
+	evl_put_thread_wchan(wchan);
+}
+
+/*
+ * Walking the PI chain deals with the following items:
+ *
+ * waiter := the thread which waits on wchan
+ * wchan := the wait channel being processed
+ * owner := the current owner of wchan
+ *
+ * In order to go fine-grained and escape ABBA situations, we hold
+ * three locks at most while walking the PI chain, keeping a
+ * consistent locking sequence among them as follows:
+ *
+ * lock(wchan)
+ *      lock(waiter, owner)
+ *
+ * Rules:
+ *
+ * 1. holding wchan->lock guarantees that wchan->owner is stable.
+ *
+ * 2. holding thread->lock guarantees that thread->wchan is stable.
+ *
+ * 3. there is no rule #3.
+ *
+ * On entry:
+ * orig_wchan->lock held + orig_waiter->lock held, irqs off.
+ */
+int evl_walk_pi_chain(struct evl_wait_channel *orig_wchan,
+		struct evl_thread *orig_waiter,
+		enum evl_walk_mode mode)
+{
+	static atomic64_t pi_serial;
+	struct evl_thread *waiter = orig_waiter, *curr = evl_current(), *owner;
+	struct evl_wait_channel *wchan = orig_wchan, *next_wchan;
+	s64 serial = atomic64_inc_return(&pi_serial);
+	bool do_reorder = false;
+	int ret = 0;
+
+	assert_hard_lock(&orig_wchan->lock);
+	assert_hard_lock(&orig_waiter->lock);
+
+	/*
+	 * The loop is always exited with two, and only two locks
+	 * held, namely: wchan->lock and waiter->lock. Also, *wchan
+	 * advances in lock-step with *waiter.
+	 *
+	 * Since we drop locks inside the PI walk, we might have
+	 * concurrent walks on the same chain from different CPUs,
+	 * contradicting each other when it comes to the priority to
+	 * set for a thread. We solve this issue as follows:
+	 *
+	 * - a unique serial number is drawn at each invocation of
+	 * this routine (pi_serial). A wait channel maintains a cached
+	 * copy of this serial number, which is updated under lock.
+	 *
+	 * - before updating the priority of the owner of a wait
+	 * channel, the serial number drawn on entry (pi_serial) is
+	 * compared to the cached value: if this value is lower than
+	 * pi_serial, the change to wchan->owner can proceed, and the
+	 * cached copy is updated into the wait channel. Otherwise,
+	 * this means that a concurrent walk started later must
+	 * override the effect of the current one, in which case the
+	 * current PI walk is aborted.
+	 *
+	 * IOW, we ensure that the latest PI walk always overrides the
+	 * effect of walks started earlier. The initial serialization
+	 * is granted by orig_wchan->lock, which is held on entry to
+	 * this routine.
+	 */
+	for (;;) {
+		owner = wchan->owner;
+		if (!owner) /* End of PI chain, we are done. */
+			break;
+
+		if (owner == orig_waiter) {
+			ret = -EDEADLK;
+			break;
+		}
+
+		/*
+		 * Lock both the waiter and owner by increasing
+		 * address order to escape the ABBA issue.  To this
+		 * end, we have to drop waiter->lock temporarily
+		 * first, so that we can re-lock it at the right place
+		 * in the sequence. In the meantime, the waiter is
+		 * prevented to go stale by holding a reference on it.
+		 */
+		evl_get_element(&waiter->element);
+		raw_spin_unlock(&waiter->lock);
+
+		evl_double_thread_lock(waiter, owner);
+		evl_put_element(&waiter->element);
+
+		/*
+		 * Multiple PI walks may happen concurrently, detect
+		 * and abort the oldest ones based on their serial
+		 * number, ensuring the latest overrides them.
+		 */
+		if (mode != evl_pi_check && serial - wchan->pi_serial < 0)
+			break;
+
+		wchan->pi_serial = serial;
+
+		/*
+		 * Make sure the waiter did not stop waiting on wchan
+		 * while unlocked (we have been holding wchan->lock
+		 * across the unlocked section, so we know that wchan
+		 * did not go stale). If the waiter is current, it is
+		 * certainly about to block on wchan, so this check
+		 * does not apply (besides, waiter->wchan does not
+		 * point at wchan yet).
+		 */
+		if (waiter != curr && waiter->wchan != wchan)
+			break;
+
+		/*
+		 * The current wait channel may need to reorder its
+		 * internal wait queue due to a priority change for
+		 * the waiter during the previous iteration.
+		 *
+		 * This also means that for a short while, a waiter
+		 * may benefit from a priority boost not yet accounted
+		 * for in the wait queue of the channel its pends on,
+		 * which is ok since there cannot be any rescheduling
+		 * on the current CPU between both events.
+		 */
+		if (do_reorder) {
+			wchan->requeue_wait(wchan, waiter);
+			do_reorder = false;
+		}
+
+		if (mode == evl_pi_adjust) {
+			/*
+			 * If priorities already match, there is no
+			 * more change to propagate downstream. Stop
+			 * the walk now.
+			 */
+			if (waiter->wprio == owner->wprio) {
+				raw_spin_unlock(&owner->lock);
+				break;
+			}
+			/* owner.cprio <- waiter.cprio */
+			evl_track_thread_policy(owner, waiter);
+			do_reorder = true;
+		} else if (mode == evl_pi_reset) {
+			/* owner.cprio <- owner.bprio */
+			if (owner->state & T_BOOST)
+				evl_track_thread_policy(owner, owner);
+			do_reorder = true;
+		}
+
+		/*
+		 * Drop the lock on the current wait channel we don't
+		 * need anymore, so that we don't risk entering an
+		 * ABBA pattern lockdep would complain about as a
+		 * result of locking owner->wchan in
+		 * evl_get_thread_wchan().
+		 */
+		raw_spin_unlock(&waiter->lock);
+		raw_spin_unlock(&wchan->lock);
+		/* Now acquire owner->wchan if non-NULL. */
+		next_wchan = evl_get_thread_wchan(owner);
+		if (!next_wchan) {
+			raw_spin_unlock(&owner->lock);
+			raw_spin_lock(&orig_wchan->lock);
+			raw_spin_lock(&orig_waiter->lock);
+			return 0;
+		}
+
+		waiter = owner;
+		wchan = next_wchan;
+	}
+
+	if (waiter != orig_waiter) {
+		raw_spin_unlock(&waiter->lock);
+		raw_spin_unlock(&wchan->lock);
+		raw_spin_lock(&orig_wchan->lock);
+		raw_spin_lock(&orig_waiter->lock);
+	}
+
+	return ret;
+}
-- 
2.30.2

